enhancements/installer/telemeter-metrics.md

```diff
@@ -0,0 +1,430 @@
+# TYPE cluster_installation_create histogram
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="15"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="20"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="25"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="30"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="35"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="40"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="45"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="50"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="55"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="60"} 1
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="+Inf"} 1
+cluster_installation_create_sum{platform="aws",result="success",version="4.2",os="linux"} 30
+cluster_installation_create_count{platform="aws",result="success",version="4.2",os="linux"} 1
+```
+The example above represents a single sample. This sample indicates that a Linux user successfully installed a 4.2 AWS cluster, which took between 30 and 35 minutes to complete installation. 
+
+###### Label Values
+* `result` - provides limited context to the outcome of an attempted install:
+    - `ProvisioningFailed`
+    - `BootstrapFailed`
+    - `APIFailed`
+    - `OperatorsDegraded`
+    - `Success`
+* `platform` - currently we deploy to 8 platforms: `aws`, `azure`, `gcp,` `libvirt`, `none`, `openstack`, `ovirt`, and `vsphere`. The list will grow. 
+* `os` - You can find all the available combinations of [Go OS and arch values](https://github.com/golang/go/blob/master/src/go/build/syslist.go) with `go tool dist list`. At this stage I assume we only want to colllect operating systems and some of the values are unlikely/impossible:
+  - `linux`
+  - `darwin` - this could be rewritten to `macos` on the installer client if we prefer
+  - `bsd` - this would encompass the `freebsd`, `netbsd`, and `openbsd` values
+  - `windows` - future proofing? 
+  - `other` - everything else not captured above
+* `version` - represents minor releases; assuming we do not backport metrics, `4.5` will be the only value if we make it into the next release and we could expect roughly 20 values over the next five years.
+* `le` - the standard Prometheus label for histogram buckets, this represents how long a command took complete. The current proposal is to have a 5 minute step starting at 15 minutes and maxing at 60.
```

Comment:
It is not clear to me how this works if we eventually decide to adjust the buckets.  Is there a reason we can't use a [guage][1] and use Prometheus ranges to aggregate and bucket entries from different installer invocations?  Otherwise, if you want to add a new `le="65"` bucket, it seems like it would be hard to initialize the counter you're feeding, because you don't know how many of the previous runs which were > 60 but less than `+Inf` should go into the `le="65"` slot.

[1]: https://prometheus.io/docs/concepts/metric_types/#gauge

Reply:
> It is not clear to me how this works if we eventually decide to adjust the buckets.

Once we set the step and range, we won't be able to adjust past data, so it's important we agree on the minimum level of granularity.   

>  Is there a reason we can't use a guage 

I'm not sure I understand this suggestion. The value of the metric will be the aggregate count of installer metrics that fall into the duration bucket specified in the label--otherwise, using the installer duration as a value, doesn't make sense as a time series or would otherwise be difficult to get a count from.   I tried to explain the difficulty in using duration as a metric value here: https://github.com/patrickdillon/metrics-faq#standard-pushgateway-basic-example If the suggestion is to still use buckets, then that is essentially a histogram.

Reply:
> Once we set the step and range, we won't be able to adjust past data, so it's important we agree on the minimum level of granularity.

It's not just the minimum, right?  We can't change our mind about buckets ever.  Or is there a change that we can make that won't break compat with the data that Prometheus has already slurped in?

> I tried to explain the difficulty in using duration as a metric value here: https://github.com/patrickdillon/metrics-faq#standard-pushgateway-basic-example ...

I had not seen this before; useful stuff.  I hadn't realized the Prometheus histograms were a client-side construct, with buckets being chosen when the metric was defined (e.g. [here][1]).  One way to address your "Each change in the graph denotes a sample pushed to the gateway. But what does this graph tell us?" might be to push a guage with "I took 17 minutes", etc., and then use [aggregation][2] with something like:

```
quantile_over_time(0.95, cluster_installation_create_seconds[7d])
```

to get the 95% quantile for installation time over the past week.  Or with a specific threshold:

```
count_over_time(cluster_installation_create_seconds[7d] > 30*60) / 
count_over_time(cluster_installation_create_seconds[7d])
```

to get the fraction of installs over 30m in the past week.  Or does that not work for some reason?

[1]: https://godoc.org/github.com/prometheus/client_golang/prometheus#HistogramOpts
[2]: https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time

Reply:
`quantile_over_time` or `count_over_time` would not be sufficient as they act on single time-series `over_time`, whereas the histogram is a representation of all runs over time. Remember that the push aggregation gateway would still be scraped by Prometheus only every x seconds, so we don't have every push event in Prometheus but only the snapshot of the counter when Prometheus sees it at scrape time.